File: services/mailsim/ui.go
package mailsim

import (
	_ "embed"
	"html/template"
	"net/http"
	"net/url"
	"strings"
)

//go:embed ui.css
var pageCSS string

//go:embed ui.js
var pageJS string

var layoutTemplate = `<!doctype html>
<html lang="en"><head><meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>{{template "title" .}}</title>
<style>` + pageCSS + `</style>
<script type="module" src="/assets/ui.js" defer></script>
</head><body><main>
<header class="top"><a class="brand" href="/">● <span>MailSim</span></a><span class="badge">LOCAL MAIL</span><span class="scope-name">{{.Scope}}</span><span class="host" id="host"></span></header>
{{template "content" .}}
<p id="notice" role="status"></p>
<footer>Captured locally. Messages are never relayed to their recipients.</footer>
</main></body></html>`

var (
	indexPage   = template.Must(template.New("layout").Parse(layoutTemplate + indexContent))
	messagePage = template.Must(template.New("layout").Parse(layoutTemplate + messageContent))
)

const indexContent = `
{{define "title"}}MailSim · Inbox{{end}}
{{define "content"}}
<div class="heading"><div><p class="eyebrow">YOUR STACK'S OUTBOX</p><h1>Every email, right here.</h1><p class="lede">Inspect invites, sign-in links and notifications as your app sends them.</p></div><span class="live" id="live-status" role="status">Live inbox</span></div>
<section class="scope-panel panel" aria-label="Inbox scope">
<div><p class="eyebrow">INBOX FOR {{.Scope}}</p><h2>One stack. Every recipient.</h2><p>Mail sent to this stack's SMTP listener lands here, whatever the recipient address. Other stacks have separate inboxes, even when they use the same email address.</p><p class="secondary">{{if .Persistent}}Captured messages survive service restarts.{{else}}This inbox is in memory and clears when MailSim restarts.{{end}}</p></div>
<div><h2>Addresses you can use</h2><p>Any email address works for capture. No mailbox setup is needed.</p><div class="address-example"><code>alex+invite@example.test</code><button type="button" data-copy="alex+invite@example.test">Copy address</button></div><p class="secondary">Receiving mail here does not create an app account. Sign up or invite that address in your app first.</p><details><summary>Connection details</summary><dl><div class="field"><dt>SMTP listener</dt><dd class="mono">{{.SMTP}}</dd></div><div class="field"><dt>Inbox</dt><dd class="mono">{{.BaseURL}}</dd></div></dl></details></div>
</section>
<section class="panel recipients-panel" aria-label="Recipient addresses"><div class="section-heading"><h2>Recipients in this inbox <span class="secondary" id="recipient-count"></span></h2><button type="button" id="all-recipients">All recipients</button></div><p class="secondary">Addresses from retained messages. Select one to filter the inbox.</p><div id="recipients" class="recipients"><p class="secondary">No addresses yet. They appear when this stack sends mail.</p></div></section>
<div class="toolbar"><label class="search">Search inbox<input id="search" type="search" placeholder="Subject, recipient or sender" autocomplete="off"></label><span id="count">{{len .Messages}} messages</span><button id="refresh" type="button">Refresh</button><button id="notify" type="button" aria-pressed="false" title="Show a desktop notification when a message arrives, even when this tab is in the background">Notify me</button><button id="clear" class="danger" type="button">Clear inbox</button></div>
<section class="panel inbox" aria-label="Caught messages"><div class="scroll"><table>
<thead><tr><th>Message</th><th>Recipient</th><th>Received</th><th>Size</th></tr></thead>
<tbody id="messages">{{range .Messages}}
<tr data-message="{{.ID}}"><td><a class="subject" href="/messages/{{.ID}}">{{if .Subject}}{{.Subject}}{{else}}(no subject){{end}}</a><div class="secondary">{{.From}}</div></td><td class="mono">{{range .To}}<div>{{.}}</div>{{end}}</td><td class="mono">{{.ReceivedAt.Format "Jan 02, 15:04:05"}}</td><td class="secondary">{{.SizeBytes}} B</td></tr>
{{end}}</tbody></table></div>
<p class="empty" id="empty" {{if .Messages}}hidden{{end}}>No messages yet. Trigger an invite or sign-in email in your app and it will appear here.</p></section>
{{end}}`

const messageContent = `
{{define "title"}}MailSim · {{.Message.Subject}}{{end}}
{{define "content"}}
<p class="crumb"><a href="/">← Inbox</a></p>
<h1 class="message-title">{{if .Message.Subject}}{{.Message.Subject}}{{else}}(no subject){{end}}</h1>
<section class="panel"><dl>
<div class="field"><dt>From</dt><dd class="mono">{{.Message.From}}</dd></div>
<div class="field"><dt>To</dt><dd class="mono">{{range .Message.To}}{{.}} {{end}}</dd></div>
<div class="field"><dt>Received</dt><dd>{{.Message.ReceivedAt.Format "02 Jan 2006, 15:04:05 MST"}}</dd></div>
<div class="field"><dt>Size</dt><dd>{{.Message.SizeBytes}} bytes</dd></div>
</dl></section>
<div class="toolbar"><div class="tabs" role="tablist" aria-label="Message format">
{{if .Message.HTML}}<button type="button" id="tab-preview" role="tab" data-view="preview" aria-controls="preview" aria-selected="true">Preview</button>{{end}}
<button type="button" id="tab-text" role="tab" data-view="text" aria-controls="text" aria-selected="{{if .Message.HTML}}false{{else}}true{{end}}">Plain text</button>
<button type="button" id="tab-headers" role="tab" data-view="headers" aria-controls="headers" aria-selected="false">Headers</button></div>
<a class="btn" href="/api/messages/{{.Message.ID}}">View JSON</a><button type="button" data-delete="{{.Message.ID}}" class="danger">Delete message</button></div>
{{if .Message.HTML}}<section id="preview" role="tabpanel" aria-labelledby="tab-preview"><iframe class="body" title="Email preview" sandbox="allow-popups allow-popups-to-escape-sandbox" src="/api/messages/{{.Message.ID}}/html"></iframe></section>{{end}}
<section class="panel" id="text" role="tabpanel" aria-labelledby="tab-text" {{if .Message.HTML}}hidden{{end}}><pre>{{if .Message.Text}}{{.Message.LinkifiedText}}{{else}}This message has no plain-text body.{{end}}</pre></section>
<section class="panel" id="headers" role="tabpanel" aria-labelledby="tab-headers" hidden><dl>{{range $key, $value := .Message.Headers}}<div class="field"><dt class="mono">{{$key}}</dt><dd class="mono">{{$value}}</dd></div>{{end}}</dl></section>
{{if .Message.Links}}<section class="panel"><h2>Links in this email</h2><ul class="links">{{range .Message.Links}}<li><a href="{{.}}" target="_blank" rel="noreferrer">{{.}}</a><button type="button" data-copy="{{.}}">Copy</button></li>{{end}}</ul></section>{{end}}
{{if .Message.Attachments}}<section class="panel"><h2>Attachments</h2>{{range .Message.Attachments}}<p>{{.Filename}} <span class="secondary">{{.ContentType}} · {{.SizeBytes}} bytes</span></p>{{end}}<p class="secondary">Attachment metadata only.</p></section>{{end}}
{{end}}`

func serveUIScript(w http.ResponseWriter, _ *http.Request) {
	setBaseHeaders(w)
	w.Header().Set("Content-Type", "text/javascript; charset=utf-8")
	w.Header().Set("Cache-Control", "no-cache")
	_, _ = w.Write([]byte(pageJS))
}

// handleUIIndex lists caught messages newest first.
func (s *Server) handleUIIndex(w http.ResponseWriter, r *http.Request) {
	if r.URL.Path != "/" {
		http.NotFound(w, r)
		return
	}
	setUIHeaders(w)
	renderPage(w, indexPage, s.pageData("Messages", s.store.List("", "")))
}

// handleUIMessage shows one message's headers, text and — via a sandboxed
// frame onto the API's own HTML endpoint — its HTML body.
func (s *Server) handleUIMessage(w http.ResponseWriter, r *http.Request) {
	msg, ok := s.store.Get(r.PathValue("id"))
	if !ok {
		setUIHeaders(w)
		http.NotFound(w, r)
		return
	}
	setUIHeaders(w)
	renderPage(w, messagePage, s.pageData("Message", msg))
}

func renderPage(w http.ResponseWriter, tmpl *template.Template, data any) {
	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	_ = tmpl.Execute(w, data)
}

func (s *Server) pageData(key string, value any) map[string]any {
	scope := "Standalone MailSim"
	if origin, err := url.Parse(s.cfg.BaseURL); err == nil && strings.HasPrefix(origin.Hostname(), "mail.") && strings.HasSuffix(origin.Hostname(), ".langwatch.localhost") {
		scope = strings.TrimSuffix(strings.TrimPrefix(origin.Hostname(), "mail."), ".langwatch.localhost")
	}
	smtp := s.cfg.SMTPAddr
	if strings.HasPrefix(smtp, ":") {
		smtp = "127.0.0.1" + smtp
	}
	return map[string]any{key: value, "Scope": scope, "SMTP": smtp, "BaseURL": s.cfg.BaseURL, "Persistent": s.cfg.DataDir != ""}
}

File: services/mailsim/ui.js
const get = (id) => document.getElementById(id);
const notice = get("notice");
get("host").textContent = location.host;
const inbox = get("messages");
const search = get("search");
let messages = [];
let busy = false;
let loaded = false;
let signature = "";
let recipient = "";
let recipientSignature = "";
// Desktop notifications. The inbox is a tab you leave open in the background
// while you work in the application that sends the mail, so the useful moment
// to be told about a message is exactly the moment this tab cannot show you
// one. Off until asked for: a page that demands notification permission on
// load is a page people deny permanently.
const NOTIFY_KEY = "mailsim:notify";
const notifyButton = get("notify");
let notifying = false;
let known = null;

function setNotifying(on) {
  notifying = on;
  if (notifyButton) {
    notifyButton.setAttribute("aria-pressed", String(on));
    notifyButton.textContent = on ? "Notifying" : "Notify me";
  }
  try {
    localStorage.setItem(NOTIFY_KEY, on ? "1" : "0");
  } catch {
    // A browser refusing storage still notifies for this session.
  }
}

// announce fires one notification per message that arrived since the last poll.
// The first poll only records what is already there — a page opened on a full
// inbox must not fire twelve notifications for mail that arrived yesterday.
function announce(list) {
  const ids = new Set(list.map((message) => message.id));
  if (known === null) {
    known = ids;
    return;
  }
  const fresh = list.filter((message) => !known.has(message.id));
  known = ids;
  if (!notifying || Notification.permission !== "granted") return;
  for (const message of fresh.slice(0, 3)) {
    const note = new Notification(message.subject || "(no subject)", {
      body: "To " + message.to.join(", "),
      tag: message.id,
    });
    note.addEventListener("click", () => {
      window.focus();
      location.assign("/messages/" + encodeURIComponent(message.id));
    });
  }
  if (fresh.length > 3) {
    new Notification(fresh.length + " messages arrived", { tag: "mailsim:batch" });
  }
}

async function api(path, options = {}) {
  const response = await fetch(path, {
    cache: "no-store",
    signal: AbortSignal.timeout(5000),
    ...options,
  });
  if (!response.ok) throw new Error("Request failed (" + response.status + "). Please retry.");
  return response.status === 204 ? null : response.json();
}

function render() {
  if (!loaded) return;
  const query = search.value.toLowerCase();
  const filtered = messages.filter(
    (message) =>
      (!recipient || message.to.some((address) => address.toLowerCase() === recipient)) &&
      [message.from, ...message.to, message.subject].join(" ").toLowerCase().includes(query),
  );
  get("count").textContent =
    filtered.length + (query || recipient ? " of " + messages.length : "") + " messages";
  get("empty").hidden = filtered.length > 0;
  get("empty").textContent =
    query || recipient
      ? "No messages match this search."
      : "No messages yet. Trigger an invite or sign-in email in your app and it will appear here.";
  get("clear").disabled = messages.length === 0;
  const next = JSON.stringify(filtered);
  if (next === signature) return;
  if (inbox.contains(document.activeElement)) return;
  const selection = String(window.getSelection());
  if (selection) return;
  signature = next;
  inbox.replaceChildren(
    ...filtered.map((message) => {
      const row = document.createElement("tr");
      row.dataset.message = message.id;
      const first = document.createElement("td");
      const link = document.createElement("a");
      link.href = "/messages/" + encodeURIComponent(message.id);
      link.className = "subject";
      link.textContent = message.subject || "(no subject)";
      const sender = document.createElement("div");
      sender.className = "secondary";
      sender.textContent = message.from;
      first.append(link, sender);
      row.append(first);
      for (const text of [
        message.to.join(", "),
        new Date(message.receivedAt).toLocaleString(),
        new Intl.NumberFormat().format(message.sizeBytes) + " B",
      ]) {
        const cell = document.createElement("td");
        cell.className = "mono";
        cell.textContent = text;
        row.append(cell);
      }
      return row;
    }),
  );
}

async function refresh() {
  // A hidden tab normally stops polling, because nobody is looking at it. With
  // notifications on somebody is waiting on it, so it keeps going — that is the
  // whole of what "notify me" buys.
  if (!inbox || busy || (document.hidden && !notifying)) return;
  busy = true;
  try {
    messages = (await api("/api/messages")).messages;
    loaded = true;
    announce(messages);
    renderRecipients();
    render();
    get("live-status").textContent = "Live · " + new Date().toLocaleTimeString();
  } catch (error) {
    get("live-status").textContent = "Disconnected · retrying";
    notice.textContent = error.message;
  } finally {
    busy = false;
  }
}

function renderRecipients() {
  const counts = new Map();
  for (const message of messages) {
    for (const address of new Set(message.to.map((value) => value.toLowerCase()))) {
      counts.set(address, (counts.get(address) || 0) + 1);
    }
  }
  const addresses = [...counts].toSorted(([a], [b]) => a.localeCompare(b));
  const next = JSON.stringify([addresses, recipient]);
  if (next === recipientSignature) return;
  recipientSignature = next;
  const focused = document.activeElement?.dataset.recipient;
  get("recipient-count").textContent = "· " + addresses.length;
  get("all-recipients").setAttribute("aria-pressed", String(!recipient));
  const nodes = addresses.map(([address, count]) => {
    const button = document.createElement("button");
    button.type = "button";
    button.dataset.recipient = address;
    button.setAttribute("aria-pressed", String(address === recipient));
    button.textContent = address + " · " + count;
    button.addEventListener("click", () => {
      recipient = address;
      renderRecipients();
      render();
    });
    return button;
  });
  if (!nodes.length) {
    const empty = document.createElement("p");
    empty.className = "secondary";
    empty.textContent = "No addresses yet. They appear when this stack sends mail.";
    nodes.push(empty);
  }
  get("recipients").replaceChildren(...nodes);
  if (focused) nodes.find((node) => node.dataset.recipient === focused)?.focus();
}

if (inbox) {
  get("all-recipients").addEventListener("click", () => {
    recipient = "";
    renderRecipients();
    render();
  });
  search.addEventListener("input", render);
  get("refresh").addEventListener("click", refresh);
  if (notifyButton) {
    if (!("Notification" in window)) {
      notifyButton.disabled = true;
      notifyButton.title = "This browser does not offer desktop notifications.";
    } else {
      let remembered = "0";
      try {
        remembered = localStorage.getItem(NOTIFY_KEY) || "0";
      } catch {
        // Storage refused; the toggle simply starts off.
      }
      setNotifying(remembered === "1" && Notification.permission === "granted");
      notifyButton.addEventListener("click", async () => {
        if (notifying) {
          setNotifying(false);
          notice.textContent = "Desktop notifications off.";
          return;
        }
        const permission =
          Notification.permission === "granted"
            ? "granted"
            : await Notification.requestPermission();
        if (permission !== "granted") {
          notice.textContent =
            "Your browser is blocking notifications for this site. Allow them in its site settings to turn this on.";
          return;
        }
        setNotifying(true);
        notice.textContent =
          "You will be told when a message arrives, even with this tab in the background.";
      });
    }
  }
  get("clear").addEventListener("click", async () => {
    if (!confirm("Clear all caught messages from this stack?")) return;
    get("clear").disabled = true;
    try {
      await api("/api/messages", { method: "DELETE" });
      notice.textContent = "Inbox cleared.";
      await refresh();
    } catch (error) {
      notice.textContent = error.message;
      get("clear").disabled = false;
    }
  });
  void refresh();
  setInterval(refresh, 2000);
  document.addEventListener("visibilitychange", refresh);
}

function selectView(tab) {
  for (const button of document.querySelectorAll("[data-view]")) {
    const active = button === tab;
    button.setAttribute("aria-selected", String(active));
    button.tabIndex = active ? 0 : -1;
    get(button.dataset.view).hidden = !active;
  }
}
const tabs = Array.from(document.querySelectorAll("[data-view]"));
if (tabs.length) selectView(tabs.find((tab) => tab.getAttribute("aria-selected") === "true"));
for (const tab of tabs) {
  tab.addEventListener("keydown", (event) => {
    if (!["ArrowLeft", "ArrowRight", "Home", "End"].includes(event.key)) return;
    event.preventDefault();
    const index = tabs.indexOf(tab);
    let next = (index + (event.key === "ArrowRight" ? 1 : -1) + tabs.length) % tabs.length;
    if (event.key === "Home") next = 0;
    if (event.key === "End") next = tabs.length - 1;
    selectView(tabs[next]);
    tabs[next].focus();
  });
}
document.addEventListener("click", async (event) => {
  const tab = event.target.closest("[data-view]");
  if (tab) selectView(tab);
  const copy = event.target.closest("[data-copy]");
  if (copy) {
    try {
      await navigator.clipboard.writeText(copy.dataset.copy);
      notice.textContent = "Copied.";
    } catch {
      notice.textContent = "Clipboard unavailable. Select the link to copy it.";
    }
  }
  const button = event.target.closest("[data-delete]");
  if (button && confirm("Delete this caught message?")) {
    button.disabled = true;
    try {
      await api("/api/messages/" + encodeURIComponent(button.dataset.delete), { method: "DELETE" });
      location.assign("/");
    } catch (error) {
      notice.textContent = error.message;
      button.disabled = false;
    }
  }
});
